OpenRA.Mods.Common/SpriteLoaders/PngSheetLoader.cs

```diff
@@ -0,0 +1,149 @@
+		}
+
+		public bool TryParseSprite(Stream s, out ISpriteFrame[] frames)
+		{
+			if (!Png.Verify(s))
+			{
+				frames = null;
+				return false;
+			}
+
+			var png = new Png(s);
+			Rectangle[] frameRegions;
+			float2[] frameOffsets;
+
+			// Prefer manual defined regions over auto sliced regions.
+			if (png.Meta.Any(meta => meta.Key.StartsWith("Frame[")))
+				RegionsFromFrames(png, out frameRegions, out frameOffsets);
+			else
+				RegionsFromSlices(png, out frameRegions, out frameOffsets);
+
+			frames = new ISpriteFrame[frameRegions.Length];
+
+			for (var i = 0; i < frames.Length; i++)
+			{
+				frames[i] = new PngSheetFrame()
+				{
+					Size = new Size(frameRegions[i].Width, frameRegions[i].Height),
+					FrameSize = new Size(frameRegions[i].Width, frameRegions[i].Height),
+					Offset = frameOffsets[i],
+					Data = new byte[frameRegions[i].Width * frameRegions[i].Height]
+				};
+
+				var frameStart = frameRegions[i].X + frameRegions[i].Y * png.Width;
+
+				for (var y = 0; y < frames[i].Size.Height; y++)
+					Array.Copy(png.Data, frameStart + y * png.Width, frames[i].Data, y * frames[i].Size.Width, frames[i].Size.Width);
+			}
+
+			return true;
+		}
+
+		void RegionsFromFrames(Png png, out Rectangle[] frameRegions, out float2[] frameOffsets)
+		{
+			var regions = new List<Rectangle>();
+			var offsets = new List<float2>();
+
+			string frame;
+			for (var i = 0; png.Meta.TryGetValue("Frame[" + i + "]", out frame); i++)
+			{
+				// Format: x,y,width,height,offsetX,offsetY
+				var coords = frame.Split(',');
+				regions.Add(new Rectangle(int.Parse(coords[0]), int.Parse(coords[1]), int.Parse(coords[2]), int.Parse(coords[3])));
+				offsets.Add(new float2(int.Parse(coords[4]), int.Parse(coords[5])));
+			}
+
+			frameRegions = regions.ToArray();
+			frameOffsets = offsets.ToArray();
+		}
+
+		void RegionsFromSlices(Png png, out Rectangle[] frameRegions, out float2[] frameOffsets)
+		{
+			Size frameSize;
+			int frameAmount;
+
+			if (png.Meta.ContainsKey("FrameSize"))
+			{
+				// If FrameSize exist, use it and...
+				var dimensions = png.Meta["FrameSize"].Split(',');
+				frameSize = new Size(int.Parse(dimensions[0]), int.Parse(dimensions[1]));
+
+				// ... either use FrameAmount or calculate how many times FrameSize fits into the image.
+				if (png.Meta.ContainsKey("FrameAmount"))
+					frameAmount = int.Parse(png.Meta["FrameAmount"]);
+				else
+					frameAmount = png.Width / frameSize.Width * png.Height / frameSize.Height;
+			}
+			else
+			{
+				if (png.Meta.ContainsKey("FrameAmount"))
+				{
+					// Otherwise, calculate the number of frames by splitting the image horizontaly by FrameAmount...
+					frameAmount = int.Parse(png.Meta["FrameAmount"]);
+					frameSize = new Size(png.Width / frameAmount, png.Height);
+				}
+				else
+				{
+					// ... or if no FrameSize and no FrameAmount is set, assume the whole image is a single frame.
+					frameAmount = 1;
```

Comment:
It would help readability if we put these two as the initial values  at the start, and then collapsed the else { if { } } into an else if { }